hare/mock_oracle.go

```diff
@@ -14,34 +17,72 @@ const (
 )
 
 type Rolacle interface {
-	Role(r uint32, sig Signature) Role
+	Role(sig Signature) Role
 }
 
 type MockHashOracle struct {
-	isLeaderTaken bool
+	clients    map[string]struct{}
+	comitySize int
+	mutex      sync.Mutex
 }
 
-func NewMockOracle() *MockHashOracle {
-	mock := &MockHashOracle{}
+// N is the expected comity size
+func NewMockHashOracle(expectedSize int, comitySize int) *MockHashOracle {
+	mock := new(MockHashOracle)
+	mock.clients = make(map[string]struct{}, expectedSize)
+	mock.comitySize = comitySize
 
 	return mock
 }
 
-func (mockOracle *MockHashOracle) Role(r uint32, proof Signature) Role {
+func (mock *MockHashOracle) Register(pubKey crypto.PublicKey) {
+	mock.mutex.Lock()
+	defer mock.mutex.Unlock()
+
+	if _, exist := mock.clients[pubKey.String()]; exist {
+		return
+	}
+
+	mock.clients[pubKey.String()] = struct{}{}
+}
+
+func (mock *MockHashOracle) Unregister(pubKey crypto.PublicKey) {
+	mock.mutex.Lock()
+	delete(mock.clients, pubKey.String())
+	mock.mutex.Unlock()
+}
+
+func (mock *MockHashOracle) calcThreshold(comitySize int) uint32 {
+	if len(mock.clients) == 0 {
```

Comment:
you should also lock when checking for the length. Consider using RWMutex instead of Mutex

Reply:
It is good to have this check to detect bug, but you also need to check that `committeeSize <= len(mock.clients)`

Reply:
Fixed.